File: native/cpp/Metrics.cpp
#include "Metrics.h"
#include <unordered_map>
#include <cmath>
#include <numeric>
#include <iostream>

using namespace Eigen;

// Function to calculate Purity
double calculatePurity(const std::vector<int>& trueLabels, const std::vector<int>& predictedLabels) {
/*
    std::cout << "True labels: ";
    for (int label : trueLabels) {
        std::cout << label << " ";
    }
    std::cout << "\nPredicted labels: ";
    for (int label : predictedLabels) {
        std::cout << label << " ";
    }
    std::cout << "\n";
*/
    std::unordered_map<int, std::unordered_map<int, int>> contingencyTable;
    for (size_t i = 0; i < trueLabels.size(); ++i) {
        contingencyTable[predictedLabels[i]][trueLabels[i]]++;
    }

    double totalCount = trueLabels.size();
    double correctCount = 0;

    // For each predicted cluster, find the true label that appears most frequently
    for (const auto& cluster : contingencyTable) {
        int maxCount = 0;
        for (const auto& label : cluster.second) {
            maxCount = std::max(maxCount, label.second);
        }
        correctCount += maxCount;
    }

    return correctCount / totalCount;
}

// Helper functions for NMI and ARI
double entropy(const std::vector<int>& labels) {
    std::unordered_map<int, int> labelCounts;
    for (int label : labels) {
        labelCounts[label]++;
    }
    double ent = 0.0;
    for (const auto& entry : labelCounts) {
        double p = static_cast<double>(entry.second) / labels.size();
        ent -= p * std::log2(p);
    }
    return ent;
}

double calculateMutualInformation(const std::vector<int>& trueLabels, const std::vector<int>& predictedLabels) {
    std::unordered_map<int, std::unordered_map<int, int>> contingencyTable;

    for (size_t i = 0; i < trueLabels.size(); ++i) {
        contingencyTable[predictedLabels[i]][trueLabels[i]]++;
    }

    double mutualInformation = 0.0;
    int N = trueLabels.size();

    for (const auto& cluster : contingencyTable) {
        int clusterSize = 0;
        for (const auto& label : cluster.second) {
            clusterSize += label.second;
        }

        for (const auto& label : cluster.second) {
            int nij = label.second;
            if (nij > 0) {
                double pij = static_cast<double>(nij) / N;
                double pi = static_cast<double>(clusterSize) / N;
                double pj = static_cast<double>(std::accumulate(trueLabels.begin(), trueLabels.end(), 0,
                        [&label](int sum, int lbl) { return sum + (lbl == label.first); })) / N;

                mutualInformation += pij * std::log2(pij / (pi * pj));
            }
        }
    }
    return mutualInformation;
}

// Function to calculate NMI
double calculateNMI(const std::vector<int>& trueLabels, const std::vector<int>& predictedLabels) {
    double hTrue = entropy(trueLabels);
    double hPred = entropy(predictedLabels);
    double mi = calculateMutualInformation(trueLabels, predictedLabels);
    return 2 * mi / (hTrue + hPred);
}

// Function to calculate ARI
double calculateARI(const std::vector<int>& trueLabels, const std::vector<int>& predictedLabels) {
    size_t N = trueLabels.size();
    std::unordered_map<int, std::unordered_map<int, int>> contingencyTable;

    for (size_t i = 0; i < N; ++i) {
        contingencyTable[predictedLabels[i]][trueLabels[i]]++;
    }

    int sumCombinations = 0, sumRowCombinations = 0, sumColCombinations = 0;
    for (const auto& cluster : contingencyTable) {
        int rowSum = 0;
        for (const auto& label : cluster.second) {
            int nij = label.second;
            sumCombinations += nij * (nij - 1) / 2;
            rowSum += nij;
        }
        sumRowCombinations += rowSum * (rowSum - 1) / 2;
    }

    for (int label : trueLabels) {
        int colSum = std::accumulate(predictedLabels.begin(), predictedLabels.end(), 0,
            [label](int sum, int predLabel) { return sum + (predLabel == label); });
        sumColCombinations += colSum * (colSum - 1) / 2;
    }

    double expectedIndex = static_cast<double>(sumRowCombinations) * sumColCombinations / (N * (N - 1) / 2);
    double maxIndex = 0.5 * (sumRowCombinations + sumColCombinations);
    return (sumCombinations - expectedIndex) / (maxIndex - expectedIndex);
}

Eigen::MatrixXi computeConfusionMatrix(const std::vector<int>& trueLabels, const std::vector<int>& predictedLabels, int numClasses) {
    Eigen::MatrixXi confusionMatrix = Eigen::MatrixXi::Zero(numClasses, numClasses);

    /*
    std::cout << "True labels: " << trueLabels.size() << std::endl;

    std::cout << "True labels: ";
    for (int label : trueLabels) {
        std::cout << label << " ";
    }
    std::cout << std::endl;

    std::cout << "Predicted labels: ";
    for (int label : predictedLabels) {
        std::cout << label << " ";
    }
    std::cout << std::endl;*/

    for (size_t i = 0; i < trueLabels.size(); ++i) {
        int trueLabel = trueLabels[i];
        int predictedLabel = predictedLabels[i];

        if (trueLabel >= 0 && trueLabel < numClasses && predictedLabel >= 0 && predictedLabel < numClasses) {
            confusionMatrix(trueLabel, predictedLabel)++;
        }
    }

    return confusionMatrix;
}

Eigen::MatrixXi findBestPermutationMatrix(const Eigen::MatrixXi& confusionMatrix) {
    int n = confusionMatrix.rows();
    std::vector<int> perm(n);
    std::vector<int> bestPerm(n);
    std::iota(perm.begin(), perm.end(), 0);
    
    int maxDiagonalSum = 0;
    
    do {
        int currentSum = 0;
        for (int i = 0; i < n; i++) {
            currentSum += confusionMatrix(i, perm[i]);
        }
        
        if (currentSum > maxDiagonalSum) {
            maxDiagonalSum = currentSum;
            bestPerm = perm;
        }
    } while (std::next_permutation(perm.begin(), perm.end()));
    
    MatrixXi permutedMatrix = Eigen::MatrixXi::Zero(n, n);

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            permutedMatrix(i, j) = confusionMatrix(i, bestPerm[j]);
        }
    }
    
    return permutedMatrix;
}

File: native/rust/algofuzz/src-nalgebra/metrics.rs
use nalgebra::DMatrix;
use std::collections::HashMap;
use std::f64::consts::LOG2_E;

pub fn calculate_purity(true_labels: &[usize], predicted_labels: &[usize]) -> f64 {
    let mut contingency_table: HashMap<usize, HashMap<usize, usize>> = HashMap::new();
    for (&true_label, &predicted_label) in true_labels.iter().zip(predicted_labels.iter()) {
        *contingency_table.entry(predicted_label).or_default().entry(true_label).or_default() += 1;
    }

    let total_count = true_labels.len();
    let mut correct_count = 0;

    for cluster in contingency_table.values() {
        let max_count = cluster.values().max().unwrap_or(&0);
        correct_count += max_count;
    }

    correct_count as f64 / total_count as f64
}

fn entropy(labels: &[usize]) -> f64 {
    let mut label_counts: HashMap<usize, usize> = HashMap::new();
    for &label in labels {
        *label_counts.entry(label).or_default() += 1;
    }

    let total_count = labels.len() as f64;
    label_counts.values().fold(0.0, |ent, &count| {
        let p = count as f64 / total_count;
        ent - p * (p * LOG2_E).ln()
    })
}

fn calculate_mutual_information(true_labels: &[usize], predicted_labels: &[usize]) -> f64 {
    let mut contingency_table: HashMap<usize, HashMap<usize, usize>> = HashMap::new();
    for (&true_label, &predicted_label) in true_labels.iter().zip(predicted_labels.iter()) {
        *contingency_table.entry(predicted_label).or_default().entry(true_label).or_default() += 1;
    }

    let total_count = true_labels.len() as f64;
    let mut mutual_information = 0.0;

    for cluster in contingency_table.values() {
        let cluster_size: usize = cluster.values().sum();
        for (&label, &nij) in cluster {
            if nij > 0 {
                let pij = nij as f64 / total_count;
                let pi = cluster_size as f64 / total_count;
                let pj = true_labels.iter().filter(|&&l| l == label).count() as f64 / total_count;
                mutual_information += pij * (pij / (pi * pj)).ln();
            }
        }
    }

    mutual_information
}

pub fn calculate_nmi(true_labels: &[usize], predicted_labels: &[usize]) -> f64 {
    let h_true = entropy(true_labels);
    let h_pred = entropy(predicted_labels);
    let mi = calculate_mutual_information(true_labels, predicted_labels);
    2.0 * mi / (h_true + h_pred)
}

pub fn calculate_ari(true_labels: &[usize], predicted_labels: &[usize]) -> f64 {
    let n = true_labels.len();
    let mut contingency_table: HashMap<usize, HashMap<usize, usize>> = HashMap::new();

    for (&true_label, &predicted_label) in true_labels.iter().zip(predicted_labels.iter()) {
        *contingency_table.entry(predicted_label).or_default().entry(true_label).or_default() += 1;
    }

    let mut sum_combinations = 0;
    let mut sum_row_combinations = 0;
    let mut sum_col_combinations = 0;

    for cluster in contingency_table.values() {
        let row_sum: usize = cluster.values().sum();
        sum_row_combinations += row_sum * (row_sum - 1) / 2;
        for &nij in cluster.values() {
            sum_combinations += nij * (nij - 1) / 2;
        }
    }

    for &label in true_labels {
        let col_sum = predicted_labels.iter().filter(|&&pred_label| pred_label == label).count();
        sum_col_combinations += col_sum * (col_sum - 1) / 2;
    }

    let expected_index = (sum_row_combinations as f64) * (sum_col_combinations as f64) / (n * (n - 1) / 2) as f64;
    let max_index = 0.5 * (sum_row_combinations + sum_col_combinations) as f64;
    (sum_combinations as f64 - expected_index) / (max_index - expected_index)
}

pub fn compute_confusion_matrix(true_labels: &[usize], predicted_labels: &[usize], num_classes: usize) -> DMatrix<usize> {
    let mut confusion_matrix = DMatrix::zeros(num_classes, num_classes);

    for (&true_label, &predicted_label) in true_labels.iter().zip(predicted_labels.iter()) {
        if true_label < num_classes && predicted_label < num_classes {
            confusion_matrix[(true_label, predicted_label)] += 1;
        }
    }

    confusion_matrix
}

pub fn find_best_permutation_matrix(confusion_matrix: &DMatrix<usize>) -> DMatrix<usize> {
    let n = confusion_matrix.nrows();
    let mut perm: Vec<usize> = (0..n).collect();
    let mut best_perm = perm.clone();
    let mut max_diagonal_sum = 0;

    loop {
        let current_sum: usize = (0..n).map(|i| confusion_matrix[(i, perm[i])]).sum();
        if current_sum > max_diagonal_sum {
            max_diagonal_sum = current_sum;
            best_perm = perm.clone();
        }
        if !next_permutation(&mut perm) {
            break;
        }
    }

    let mut permuted_matrix = DMatrix::zeros(n, n);
    for i in 0..n {
        for j in 0..n {
            permuted_matrix[(i, j)] = confusion_matrix[(i, best_perm[j])];
        }
    }

    permuted_matrix
}

pub fn next_permutation<T: Ord>(slice: &mut [T]) -> bool {
    if slice.len() < 2 {
        return false;
    }

    let mut i = slice.len() - 1;
    while i > 0 && slice[i - 1] >= slice[i] {
        i -= 1;
    }

    if i == 0 {
        slice.reverse();
        return false;
    }

    let mut j = slice.len() - 1;
    while slice[j] <= slice[i - 1] {
        j -= 1;
    }

    slice.swap(i - 1, j);
    slice[i..].reverse();
    true
}